snapcraft/tests/test_plugin_gulp.py

```diff
@@ -0,0 +1,147 @@
+
+        self.assertFalse(self.run_mock.called, 'run() was called')
+        self.tar_mock.assert_has_calls([
+            mock.call(
+                nodejs.get_nodejs_release(plugin.options.node_engine),
+                path.join(os.path.abspath('.'), 'parts', 'test-part', 'npm')),
+            mock.call().download()])
+
+    def test_build(self):
+        self.skipTest('I cannot mock.patch.dict')
+
+        class Options:
+            source = '.'
+            gulp_tasks = []
+            node_engine = '4'
+
+        plugin = gulp.GulpPlugin('test-part', Options(), self.project_options)
+
+        os.makedirs(plugin.sourcedir)
+        open(os.path.join(plugin.sourcedir, 'package.json'), 'w').close()
+
+        with mock.patch.dict(os.environ, {'PATH': 'bin'}):
+            plugin.build()
+
+        self.run_mock.assert_has_calls([
+            mock.call(['npm', 'install', '-g', 'gulp-cli'],
+                      cwd=plugin.builddir, env={})])
+        self.tar_mock.assert_has_calls([
+            mock.call(
+                nodejs.get_nodejs_release(plugin.options.node_engine),
+                path.join(os.path.abspath('.'), 'parts', 'test-part', 'npm')),
+            mock.call().provision(
+                plugin._npm_dir, clean_target=False, keep_tarball=True)])
+
+    @mock.patch('platform.machine')
+    def test_unsupported_arch_raises_exception(self, machine_mock):
+        machine_mock.return_value = 'fantasy-arch'
+
+        class Options:
+            source = None
+            gulp_tasks = []
+            node_engine = '4'
+
+        with self.assertRaises(EnvironmentError) as raised:
+            gulp.GulpPlugin('test-part', Options(), self.project_options)
+
+        self.assertEqual(raised.exception.__str__(),
+                         'architecture not supported (fantasy-arch)')
+
+    def test_schema(self):
+        self.maxDiff = None
+        plugin_schema = {
+            '$schema': 'http://json-schema.org/draft-04/schema#',
+            'additionalProperties': False,
+            'properties': {
+                'gulp-tasks': {'default': [],
+                               'items': {'type': 'string'},
+                               'minitems': 1,
+                               'type': 'array',
+                               'uniqueItems': True},
+                'node-engine': {'default': '4.4.4', 'type': 'string'},
+                'source': {'type': 'string'},
+                'source-branch': {'default': '', 'type': 'string'},
+                'source-subdir': {'default': None, 'type': 'string'},
+                'source-tag': {'default': '', 'type:': 'string'},
+                'source-type': {'default': '', 'type': 'string'}},
+            'pull-properties': ['source', 'source-type', 'source-branch',
+                                'source-tag', 'source-subdir', 'node-engine'],
+            'build-properties': ['gulp-tasks'],
+            'required': ['gulp-tasks'],
+            'type': 'object'}
```

Comment:
A dump like this makes the tests easier to write, but harder to figure out when the test fails, so I tend to like helpful assertions.